autofaiss/indices/distributed.py

```diff
@@ -336,3 +342,315 @@ def run(
+    use_gpu: bool = False,
+    metric_type: str = "ip",
+    make_direct_map: bool = False,
+    should_be_memory_mappable: bool = False,
+    temp_root_dir: str = "hdfs://root/tmp/distributed_autofaiss_indices",
+) -> List[Optional[Dict[str, str]]]:
+    """
+    Create partitioned indexes from a list of parquet partitions,
+    i.e. create and train one index per parquet partition
+    """
+
+    def _create_small_indexes(embedding_root_dirs: List[str]) -> List[Optional[Dict[str, str]]]:
+        rdd = ss.sparkContext.parallelize(embedding_root_dirs, len(embedding_root_dirs))
+        return rdd.map(
+            lambda embedding_root_dir: create_small_index(
+                embedding_root_dir=embedding_root_dir,
+                output_root_dir=output_root_dir,
+                id_columns=id_columns,
+                should_be_memory_mappable=should_be_memory_mappable,
+                max_index_query_time_ms=max_index_query_time_ms,
+                max_index_memory_usage=max_index_memory_usage,
+                min_nearest_neighbors_to_retrieve=min_nearest_neighbors_to_retrieve,
+                embedding_column_name=embedding_column_name,
+                current_memory_available=current_memory_available,
+                use_gpu=use_gpu,
+                metric_type=metric_type,
+                nb_cores=nb_cores,
+                make_direct_map=make_direct_map,
+            )[1]
+        ).collect()
+
+    ss = _get_pyspark_active_session()
+
+    create_big_index_fn = partial(
+        create_big_index,
+        ss=ss,
+        output_root_dir=output_root_dir,
+        id_columns=id_columns,
+        should_be_memory_mappable=should_be_memory_mappable,
+        max_index_query_time_ms=max_index_query_time_ms,
+        max_index_memory_usage=max_index_memory_usage,
+        min_nearest_neighbors_to_retrieve=min_nearest_neighbors_to_retrieve,
+        embedding_column_name=embedding_column_name,
+        current_memory_available=current_memory_available,
+        nb_cores=nb_cores,
+        use_gpu=use_gpu,
+        metric_type=metric_type,
+        nb_splits_per_big_index=nb_splits_per_big_index,
+        make_direct_map=make_direct_map,
+        temp_root_dir=temp_root_dir,
+    )
+
+    # Compute number of embeddings for each partition
+    rdd = ss.sparkContext.parallelize(partitions, len(partitions))
+    partition_sizes = rdd.map(
+        lambda partition: (
+            partition,
+            EmbeddingReader(partition, file_format="parquet", embedding_column=embedding_column_name).count,
+        )
+    ).collect()
+
+    # Group partitions in two categories, small and big indexes
+    small_partitions = []
+    big_partitions = []
+    for partition, size in partition_sizes:
+        if size < big_index_threshold:
+            small_partitions.append(partition)
+        else:
+            big_partitions.append(partition)
+
+    # Create small and big indexes
+    all_metrics = []
+    with ThreadPool(256) as p:
```

Comment:
FAR: We might want even more than 256 threads if there are more than 256 tasks available and that we only work with small indices